internal/core/services/autoscaler/autoscaler.go

```diff
@@ -0,0 +1,87 @@
+// MIT License
+//
+// Copyright (c) 2021 TFG Co
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package autoscaler
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/topfreegames/maestro/internal/core/entities"
+	"github.com/topfreegames/maestro/internal/core/entities/autoscaling"
+	autoscalerPorts "github.com/topfreegames/maestro/internal/core/ports/autoscaler"
+)
+
+// PolicyFactory is a type that corelates a policy type with an autoscaling policy.
+type PolicyFactory map[autoscaling.PolicyType]autoscalerPorts.Policy
+
+// Autoscaler is a service that holds dependencies to execute autoscaling feature.
+type Autoscaler struct {
+	policyFactory PolicyFactory
+}
+
+// NewAutoscaler returns a new instance of autoscaler.
+func NewAutoscaler(policyFactory PolicyFactory) *Autoscaler {
+	autoscaler := &Autoscaler{
+		policyFactory: policyFactory,
+	}
+
+	return autoscaler
+}
+
+// CalculateDesiredNumberOfRooms return the number of rooms that a Scheduler should have based on its policy or error if it can calculate.
+func (a *Autoscaler) CalculateDesiredNumberOfRooms(ctx context.Context, scheduler *entities.Scheduler) (int, error) {
+	if scheduler.Autoscaling == nil {
+		return -1, errors.New("Scheduler does not have autoscaling struct")
+	}
+
+	if _, ok := a.policyFactory[scheduler.Autoscaling.Policy.Type]; !ok {
+		return -1, fmt.Errorf("Error finding policy to scheduler %s", scheduler.Name)
+	}
+
+	policy := a.policyFactory[scheduler.Autoscaling.Policy.Type]
+
+	currentState, err := policy.CurrentStateBuilder(ctx, scheduler)
+	if err != nil {
+		return -1, fmt.Errorf("Error fetching current state to scheduler %s: %w", scheduler.Name, err)
+	}
+
+	desiredNumberOfRooms, err := policy.CalculateDesiredNumberOfRooms(scheduler.Autoscaling.Policy.Parameters, currentState)
+	if err != nil {
+		return -1, fmt.Errorf("Error calculating the desired number of rooms to scheduler %s: %w", scheduler.Name, err)
```

Comment:
Very tiny detail: the pattern Maestro has now consists of non-capitalized errors (like all of these errors start with a Capitalized letter)